gcloud-java-storage/src/main/java/com/google/gcloud/storage/testing/FakeStorageRpc.java

```diff
@@ -0,0 +1,261 @@
+
+  @Override
+  public String open(StorageObject object, Map<Option, ?> options) throws StorageException {
+    String key = fullname(object);
+    boolean mustNotExist = false;
+    for (Option option : options.keySet()) {
+      if (option instanceof StorageRpc.Option) {
+        // this is a bit of a hack, since we don't implement generations.
+        if ((StorageRpc.Option) option == Option.IF_GENERATION_MATCH
+            && ((Long) options.get(option)).longValue() == 0L) {
+          mustNotExist = true;
+        }
+      }
+    }
+    if (mustNotExist && stuff.containsKey(key)) {
+      throw new StorageException(new FileAlreadyExistsException(key));
+    }
+    stuff.put(key, object);
+
+    return fullname(object);
+  }
+
+  @Override
+  public void write(String uploadId, byte[] toWrite, int toWriteOffset, long destOffset,
+                    int length, boolean last) throws StorageException {
+    // this may have a lot more allocations than ideal, but it'll work.
+    byte[] bytes;
+    if (contents.containsKey(uploadId)) {
+      bytes = contents.get(uploadId);
+      if (bytes.length < length + destOffset) {
+        bytes = new byte[(int) (length + destOffset)];
+      }
+    } else {
+      bytes = new byte[(int) (length + destOffset)];
+    }
+    System.arraycopy(toWrite, toWriteOffset, bytes, (int) destOffset, length);
+    contents.put(uploadId, bytes);
+  }
+
+  @Override
+  public RewriteResponse openRewrite(RewriteRequest rewriteRequest) throws StorageException {
+    String sourceKey = fullname(rewriteRequest.source);
+    // a little hackish, just good enough for the tests to work.
+    if (!contents.containsKey(sourceKey)) {
+      throw new StorageException(404, "File not found: " + sourceKey);
+    }
+
+    boolean mustNotExist = false;
+    for (Option option : rewriteRequest.targetOptions.keySet()) {
+      if (option instanceof StorageRpc.Option) {
+        // this is a bit of a hack, since we don't implement generations.
+        if ((StorageRpc.Option) option == Option.IF_GENERATION_MATCH
+            && ((Long) rewriteRequest.targetOptions.get(option)).longValue() == 0L) {
+          mustNotExist = true;
+        }
+      }
+    }
+
+    String destKey = fullname(rewriteRequest.target);
+    if (mustNotExist && contents.containsKey(destKey)) {
+      throw new StorageException(new FileAlreadyExistsException(destKey));
+    }
+
+    stuff.put(destKey, rewriteRequest.target);
+
+    byte[] data = contents.get(sourceKey);
+    contents.put(destKey, Arrays.copyOf(data, data.length));
+    return new RewriteResponse(rewriteRequest, rewriteRequest.target, data.length, true,
+        "rewriteToken goes here", data.length);
+  }
+
+  @Override
+  public RewriteResponse continueRewrite(RewriteResponse previousResponse) throws StorageException {
+    throw new UnsupportedOperationException();
+  }
+
+  private String fullname(StorageObject so) {
+    return (so.getBucket() + "/" + so.getName()).replaceAll("//", "/");
```

Comment:
This is incorrect. Cloud Storage is a flat namespace. It's perfectly happy to store empty path components. They are not canonicalized.

    jart@compy:~$ echo hello >hello.txt
    jart@compy:~$ gsutil cp hello.txt gs://bucket/foo//bar
    Copying file://hello.txt [Content-Type=text/plain]...
    Uploading   gs://bucket/foo//bar:      6 B/6 B    
    jart@compy:~$ gsutil cat gs://bucket/foo//bar
    hello
    jart@compy:~$ gsutil cat gs://bucket/foo/bar
    CommandException: No URLs matched gs://bucket/foo/bar

There is actually a setting in `CloudStorageConfiguration` called `permitEmptyPathComponents` which controls this behavior. However this class does not need to be concerned about that. It should just accept the path that it's given.